#Ajuste de redes neuronales
###Karla González Sánchez | A01541526



In [ ]:
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPRegressor, MLPClassifier
from sklearn.model_selection import cross_val_predict, GridSearchCV
from sklearn.metrics import mean_squared_error, mean_absolute_error, classification_report

#Ejercicio 1

Mi matrícula es: **A01541526**
Por lo tanto, voy a trabajar con las variables independientes **M, W, S y P.**

Y como variable dependiente MR.
  * **MR** = asesinatos por cada 100000 habitantes

  *  **M** = porcentaje de áreas metropolitanas
  *   **W** = porcentaje de gente blanca
  *   **P** = porcentaje con ingresos por debajo del nivel de pobreza
  *   **S** = porcentaje de familias con solo un miembro adulto como tutor

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Aprendizaje/crime_data.csv')

###1. Evalúa con validación cruzada un modelo pereceptrón multicapa para las variables que se te asignaron para este ejercicio.

In [ ]:
#Variables predictoras
X = df[['M', 'W', 'S', 'P']].values

#Variable de respuesta
y = df['MR'].values

In [ ]:
#Inicialización  del modelo de perceptrón multicapa
mlp = MLPRegressor(hidden_layer_sizes = (100, 100), max_iter = 10000)

In [ ]:
#Validación cruzada y obtener predicciones
y_pred = cross_val_predict(mlp, X, y, cv = 5)

In [ ]:
#MSE y MAE para evaluar el rendimiento del modelo
mse = mean_squared_error(y, y_pred)
mae = mean_absolute_error(y, y_pred)

print("MSE:", mse)
print("MAE:", mae)

MSE: 116.61234198011965
MAE: 3.390833707959911


###2. Agrega al conjunto de datos columnas que representen los cuadrados de las variables predictoras (por ejemplo, $M^2$, $W^2$), así como los productos entre pares de variables (por ejemplo, $PxS$, $MxW$). Evalúa un modelo perceptrón multicapa para este nuevo conjunto de datos.

In [ ]:
df['M^2'] = df['M'] ** 2
df['W^2'] = df['W'] ** 2
df['S^2'] = df['S'] ** 2
df['P^2'] = df['P'] ** 2

df['PxS'] = df['P'] * df['S']
df['MxW'] = df['M'] * df['W']
df.head()

State    VR    MR     M     W     H     P     S      M^2      W^2     S^2  \
0    AK   761   9.0  41.8  75.2  86.6   9.1  14.3  1747.24  5655.04  204.49   
1    AL   780  11.6  67.4  73.5  66.9  17.4  11.5  4542.76  5402.25  132.25   
2    AR   593  10.2  44.7  82.9  66.3  20.0  10.7  1998.09  6872.41  114.49   
3    AZ   715   8.6  84.7  88.6  78.7  15.4  12.1  7174.09  7849.96  146.41   
4    CA  1078  13.1  96.7  79.3  76.2  18.2  12.5  9350.89  6288.49  156.25   

      P^2     PxS      MxW  
0   82.81  130.13  3143.36  
1  302.76  200.10  4953.90  
2  400.00  214.00  3705.63  
3  237.16  186.34  7504.42  
4  331.24  227.50  7668.31

In [ ]:
#Variables independientes
X = df[['M', 'W', 'S', 'P', 'M^2', 'W^2', 'S^2', 'P^2', 'PxS', 'MxW']]

#Variable dependiente
y = df['MR']

In [ ]:
#Inicialización  del modelo de perceptrón multicapa
mlp = MLPRegressor(hidden_layer_sizes = (100, 100), max_iter = 10000)

In [ ]:
#Validación cruzada y obtener predicciones
y_pred = cross_val_predict(mlp, X, y, cv = 5)

In [ ]:
#MSE y MAE para evaluar el rendimiento del modelo
mse = mean_squared_error(y, y_pred)
mae = mean_absolute_error(y, y_pred)
print("MSE:", mse)
print("MAE:", mae)

MSE: 827.6543999567155
MAE: 14.647349219757325


###3. Viendo los resultados de regresión, desarrolla una conclusión sobre los siguientes puntos:

**1. ¿Consideras que el modelo perceptrón multicapa es efectivo para modelar los datos del problema? ¿Por qué?**

El modelo de perceptrón multicapa sí lo podemos considerar efectivo para modelar los datos del problema utilizando las variables predictoras originales. En el cual obtuvimos un MSE de 116.612 y un MAE de 3.391. Indicando un buen rendimiento en la predicción, ya que los errores son relativamente bajos.

Por otro lado, en el segundo modelo, donde se agregó cuadrados y productos entre varibles, no mejoró el rendimiento del modelo, pues tenemos valores de error mucho más altos. Por lo tanto, este cambio, probablemente trajo ruido innecesario al modelo.

**2. ¿Qué modelo es mejor para los datos de criminalidad, el lineal o el perceptrón multicapa? ¿Por qué?**

En ejercicios anteriores donde obtuvimos el modelo linael, se tienen las siguientes métricas:

* $R^2$ de 0.792

* MSE de 47913.47

* MAE de 206.37

El modelo perceptrón multicapa es más efectivo que el modelo lineal para los datos de criminalidad, especialmente cuando se utilizan solo las variables originales.

#Ejercicio 2

Mi matrícula es: **A01541526** Por lo tanto, voy a trabajar con el conjunto de datos **M_2.txt**

En este ejercicio se trabajará con el conjunto de datos M_2.txt. En estos archivos viene un experimento en el que se midió actividad muscular con la técnica de la Electromiografía en el brazo derecho de varios participantes cuando éstos realizaban un movimiento con la mano entre siete posible (Flexionar hacia arriba, Flexionar hacia abajo, Cerrar la mano, Estirar la mano, Abrir la mano, Coger un objeto, No moverse). Al igual que en el ejercicio anterior, los datos se cargan con la función loadtxt de numpy.

La primera columna corresponde a la clase (1, 2, 3, 4, 5, 6, y 7), la segunda columna se ignora, y el resto de las columnas indican las variables que se calcularon de la respuesta muscular.

In [ ]:
data = np.loadtxt('/content/drive/MyDrive/Colab Notebooks/Aprendizaje/M_2.txt', delimiter='\t', usecols=range(632))

###1. Evalúa un modelo perceptrón multicapa con validación cruzada utilizando al menos 5 capas de 20 neuronas.


In [ ]:
#Ignoraramos las primeras dos columnas, una por ser la clase y otra por ser irrelevante
X = data[:, 2:]
#Clase
y = data[:, 0]

In [ ]:
mlp = MLPClassifier(max_iter=10000, random_state=42)

#1 a 5 capas de 20 neuronas
param_grid = {
    'hidden_layer_sizes': [(20,) * i for i in range(1, 6)],
}

#Búsqueda de hiperparámetros con validación cruzada
clf = GridSearchCV(mlp, param_grid, cv=5)
clf.fit(X, y)

best_model = clf.best_estimator_

#Validación cruzada
y_pred = cross_val_predict(best_model, X, y, cv=5)

print(classification_report(y, y_pred))

              precision    recall  f1-score   support

         1.0       0.90      0.89      0.89        90
         2.0       0.71      0.81      0.76        90
         3.0       0.95      0.92      0.94        90
         4.0       0.92      0.92      0.92        90
         5.0       0.76      0.74      0.75        90
         6.0       0.80      0.70      0.75        90
         7.0       0.89      0.93      0.91        90

    accuracy                           0.85       630
   macro avg       0.85      0.85      0.85       630
weighted avg       0.85      0.85      0.85       630



###2. Evalúa un modelo perceptrón multicapa con validación cruzada, pero encontrando el número óptimo de capas y neuronas de la red.


In [ ]:
mlp = MLPClassifier(max_iter=10000, random_state=42)

#Probar combinaciones de 1 a 10 capas y de 1 a 10 neuronas en cada capa
param_grid = {
    'hidden_layer_sizes': [(i, j) for i in range(1, 11) for j in range(1, 11)],
}

#Búsqueda de hiperparámetros con validación cruzada
clf = GridSearchCV(mlp, param_grid, cv=5)
clf.fit(X, y)

best_model = clf.best_estimator_

#Validación cruzada
y_pred = cross_val_predict(best_model, X, y, cv=5)

print(classification_report(y, y_pred))

print("Mejor configuración de capas y neuronas:", best_model.hidden_layer_sizes)

              precision    recall  f1-score   support

         1.0       0.91      0.89      0.90        90
         2.0       0.73      0.78      0.75        90
         3.0       0.94      0.91      0.93        90
         4.0       0.85      0.86      0.85        90
         5.0       0.76      0.80      0.78        90
         6.0       0.79      0.72      0.76        90
         7.0       0.88      0.89      0.88        90

    accuracy                           0.83       630
   macro avg       0.84      0.83      0.84       630
weighted avg       0.84      0.83      0.84       630

Mejor configuración de capas y neuronas: (10, 8)


###3. Prepara el modelo perceptrón multicapa:
1. Obten los hiperparámetros óptimos de capas y neuronas de la red.
2. Con los hiperparámetros óptimos, ajusta el modelo con todos los datos.


In [ ]:
mlp = MLPClassifier(max_iter=10000, random_state=42)

#Probar de 1 a 20 neuronas en una sola capa oculta
param_grid = {
    'hidden_layer_sizes': [(i,) for i in range(1, 21)],
}

#Búsqueda de hiperparámetros con validación cruzada
clf = GridSearchCV(mlp, param_grid, cv=5)
clf.fit(X, y)

#Mejor modelo con los hiperparámetros óptimos
best_model = clf.best_estimator_

#Obtener las predicciones del modelo ajustado
y_pred = best_model.predict(X)

print(classification_report(y, y_pred))

print("Mejor configuración de capas y neuronas:", best_model.hidden_layer_sizes)

              precision    recall  f1-score   support

         1.0       1.00      0.99      0.99        90
         2.0       0.99      0.98      0.98        90
         3.0       1.00      1.00      1.00        90
         4.0       1.00      0.99      0.99        90
         5.0       0.97      0.99      0.98        90
         6.0       0.99      0.98      0.98        90
         7.0       0.98      1.00      0.99        90

    accuracy                           0.99       630
   macro avg       0.99      0.99      0.99       630
weighted avg       0.99      0.99      0.99       630

Mejor configuración de capas y neuronas: (16,)


In [ ]:
# Ajustar el modelo con todos los datos
best_model.fit(X, y)

MLPClassifier(hidden_layer_sizes=(16,), max_iter=10000, random_state=42)

###4. Contesta lo siguientes:
**1. ¿Observas alguna mejora importante al optimizar el tamaño de la red? ¿Es el resultado que esperabas? Argumenta tu respuesta.**

Sí, se observa una mejora en el rendimiento del modelo en comparación con el modelo de referencia sin optimización.

En el primer caso, evaluamos un modelo con 5 capas ocultas de 20 neuronas cada una. Obtenemos un rendimiento decente con una precisión de 0.85 y una puntuación F1 promedio de 0.85.

En el segundo caso, realizamos una búsqueda más exhaustiva probando todas las combinaciones posibles de capas y neuronas en el rango de 1 a 10. Obtenemos un rendimiento similar al primer caso, con una precisión promedio de 0.83 y una puntuación F1 promedio de 0.84. Aunque el rendimiento no mejora significativamente en comparación con el primer caso, el modelo se ajustó mejor a los datos.

En el tercer caso, ajustamos el modelo con una sola capa oculta de 16 neuronas, que fueron los hiperparámetros óptimos encontrados. El rendimiento del modelo mejoró considerablemente, con una precisión promedio de 0.99 y una puntuación F1 promedio de 0.99.

No me esperaba que salieran resultados tan altos en la optimización del modelo, lo cual es bueno, pero hay que tener en cuenta que existe riesgo de sobreajuste.

**2. ¿Qué inconvenientes hay al encontrar el tamaño óptimo de la red? ¿Por qué?**

Puede ser intensivo en recursos computacionales, como la capacidad de procesamiento y tiempo de ejecución, debido al entrenamiento y evaluación de múltiples configuraciones; también el riesgo de sobreajuste, capturando ruido y detalles irrelevantes, afectando el rendimiento del modelo.